DOCSP-41145: Aggregation landing page (#30)
mongodb/docs-kotlin-sync: source/aggregation.txt

.. _kotlin-sync-aggregation:

====================================
Transform Your Data with Aggregation

> ====================================

Reviewer: Q: I know this is part of the TOC we are adhering to, but would it make sense to just call this page "Transform Your Data"?

Author: I think it's helpful to leave the Aggregation in the title here since it puts a name on the subject of the page tree. "Transform Your Data" is helpful for those unfamiliar with driver concepts, while the "Aggregation" part of the name is helpful for those who are familiar.

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: code example, transform, computed, pipeline
   :description: Learn how to use the Kotlin Sync driver to perform aggregation operations.

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. .. toctree::
..    :titlesonly:
..    :maxdepth: 1

..    /aggregation/aggregation-tutorials

Overview
--------

In this guide, you can learn how to use the {+driver-short+} to perform
**aggregation operations**.

> Aggregation operations process data in your MongoDB collections and
> return computed results. The MongoDB Aggregation framework, which is
> part of the Query API, is modeled on the concept of data processing
> pipelines. Documents enter a pipeline that contains one or more stages,
> and this pipeline transforms the documents into an aggregated result.

Reviewer: ```suggestion
You can use aggregation operations to process data in your MongoDB collections and
return computed results. The MongoDB Aggregation framework, which is
part of the Query API, is modeled on the concept of a data processing
pipeline. Documents enter a pipeline that contains one or more stages,
and each stage transforms the documents to output a final aggregated result.
```

> An aggregation operation is similar to a car factory. A car factory has
> an assembly line, which contains assembly stations with specialized
> tools to do specific jobs, like drills and welders. Raw parts enter the
> factory, and then the assembly line transforms and assembles them into a
> finished product.

Reviewer: ```suggestion
You can think of an aggregation operation as similar to a car factory. A car factory has
an assembly line, which contains assembly stations with specialized
tools to do specific jobs, like drills and welders. Raw parts enter the
factory, and then the assembly line transforms and assembles them into a
finished product.
```

The **aggregation pipeline** is the assembly line, **aggregation stages** are the
assembly stations, and **operator expressions** are the
specialized tools.

> Aggregation Versus Find Operations
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Compare Aggregation and Find Operations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

You can use find operations to perform the following actions:

- Select which documents to return
- Select which fields to return
- Sort the results

You can use aggregation operations to perform the following actions:

- Perform find operations
- Rename fields
- Calculate fields
- Summarize data
- Group values

Limitations
~~~~~~~~~~~

> Keep the following limitations in mind when using aggregation operations:

Reviewer: ```suggestion
The following limitations apply when using aggregation operations:
```

- Returned documents must not violate the
  :manual:`BSON document size limit </reference/limits/#mongodb-limit-BSON-Document-Size>`
  of 16 megabytes.

> - Pipeline stages have a memory limit of 100 megabytes by default. You can exceed this
>   limit by using the ``allowDiskUse`` method of the
>   ``AggregateIterable`` type. 

Reviewer: ```suggestion
- Pipeline stages have a memory limit of 100 megabytes by default. You can exceed this
  limit by using the ``allowDiskUse()`` method from the ``AggregateIterable`` class. 
```

.. important:: $graphLookup exception

>    The :manual:`$graphLookup
>    </reference/operator/aggregation/graphLookup/>` stage has a strict
>    memory limit of 100 megabytes and ignores the ``allowDiskUse`` parameter.

Reviewer: ```suggestion
   The :manual:`$graphLookup
   </reference/operator/aggregation/graphLookup/>` stage has a strict
   memory limit of 100 megabytes and ignores the ``allowDiskUse`` option.
```

Aggregation Example
-------------------

The examples in this section use the ``restaurants`` collection in the ``sample_restaurants``
database from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
free MongoDB Atlas cluster and load the sample datasets, see the
:atlas:`Get Started with Atlas </getting-started>` guide.

The following {+language+} data class models the documents in this collection:

.. literalinclude:: /includes/aggregation/aggregation.kt
   :start-after: start-data-class
   :end-before: end-data-class
   :language: kotlin
   :copyable:

Build and Execute an Aggregation Pipeline
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To perform an aggregation, pass a list of aggregation stages to the
> ``collection.aggregate()`` method.

Reviewer: ```suggestion
To perform an aggregation on the documents in a collection, pass a list of aggregation stages to the
``aggregate()`` method.
```

> The following code example produces a count of the number of bakeries in each borough
> of New York. To do so, it uses an aggregation pipeline with the following stages:

Reviewer: ```suggestion
This example outputs a count of the number of bakeries in each borough of New York City.
The following code creates an aggregation pipeline that contains the following stages:
```

> - A :manual:`$match </reference/operator/aggregation/match/>` stage to filter for documents
>   whose ``cuisine`` field contains the value ``"Bakery"``.

Reviewer: ```suggestion
- A :manual:`$match </reference/operator/aggregation/match/>` stage to filter for documents
  in which the value of the ``cuisine`` field is ``"Bakery"``.
```

> - A :manual:`$group </reference/operator/aggregation/group/>` stage to group the matching
>   documents by the ``borough`` field, accumulating a count of documents for each distinct
>   value.

Reviewer: ```suggestion
- A :manual:`$group </reference/operator/aggregation/group/>` stage to group the matching
  documents by the ``borough`` field, producing a count of documents for each distinct
  value of that field.
```

> 

Reviewer: S: add a note here (can be commented out as a TODO) that says the following example uses Aggregates builders.

.. io-code-block::

   .. input:: /includes/aggregation/aggregation.kt
      :language: kotlin
      :start-after: start-aggregation-pipeline
      :end-before: end-aggregation-pipeline
      :dedent:

   .. output::
      :visible: false

      Document{{_id=Bronx, count=71}}
      Document{{_id=Manhattan, count=221}}
      Document{{_id=Brooklyn, count=173}}
      Document{{_id=Queens, count=204}}
      Document{{_id=Staten Island, count=20}}
      Document{{_id=Missing, count=2}}

Explain an Aggregation
~~~~~~~~~~~~~~~~~~~~~~

> To view information about how MongoDB executes your operation, you can
> instruct MongoDB to **explain** it. When MongoDB explains an operation, it returns
> **execution plans** and performance statistics. An execution
> plan is a potential way MongoDB can complete an operation.
> When you instruct MongoDB to explain an operation, it returns both the
> plan MongoDB executed and any rejected execution plans.

Reviewer: ```suggestion
To view information about how MongoDB executes your operation, you can
include the ``$explain`` aggregation stage in your pipeline. When MongoDB explains an operation,
it returns **execution plans** and performance statistics. An execution
plan is a potential way MongoDB can complete an operation.
When you instruct MongoDB to explain an operation, it returns the
plan MongoDB selected for the operation and any rejected execution plans.
```

> The following code example runs the preceding aggregation example and prints the returned
> explanation:

Reviewer: ```suggestion
The following code example runs the same aggregation shown in the preceding secction
and adds the ``$explain`` stage to output the operation details:
```

.. io-code-block::

   .. input:: /includes/aggregation/aggregation.kt
      :language: kotlin
      :start-after: start-aggregation-explain
      :end-before: end-aggregation-explain
      :dedent:

   .. output::
      :visible: false
    
      {
        "explain": {
          "aggregate": "restaurants",
          "pipeline": [
            {
              "$match": {
                "cuisine": "Bakery"
              }
            },
            {
              "$group": {
                "_id": "$borough",
                "count": {
                  "$sum": 1
                }
              }
            }
          ],
          "cursor": {}
        },
        ...

>       }

Reviewer: S: it would be helpful if this output included the info about the execution plans etc

Additional Information
----------------------

> To view a full list of expression operators, see :manual:`Aggregation
> Operators. </reference/operator/aggregation/>`

Reviewer: ```suggestion
To view a full list of expression operators, see :manual:`Aggregation
Operators </reference/operator/aggregation/>` in the (source constant for server manual)
```

Applies to all entries here

To learn about assembling an aggregation pipeline and view examples, see
:manual:`Aggregation Pipeline. </core/aggregation-pipeline/>`

To learn more about creating pipeline stages, see :manual:`Aggregation
Stages. </reference/operator/aggregation-pipeline/>`

To learn more about explaining MongoDB operations, see
:manual:`Explain Output </reference/explain-results/>` and
:manual:`Query Plans. </core/query-plans/>`

.. Aggregation Tutorials
.. ~~~~~~~~~~~~~~~~~~~~~

.. To view step-by-step explanations of common aggregation tasks, see
.. :ref:`kotlin-sync-aggregation-tutorials`.

API Documentation
~~~~~~~~~~~~~~~~~

For more information about executing aggregation operations with the {+driver-short+},
see the following API documentation:

- `aggregate() <{+api+}/com.mongodb.kotlin.client/-mongo-collection/aggregate.html>`__

> - `AggregateIterable <{+api+}/com.mongodb.kotlin.client/-aggregate-iterable/index.html>`__

Author: Note for reviewer: these links are broken at the moment.

Rea has a change that has yet to be merged that is changes the `{+api+}` source constant. This will work once it's merged.
